_deploy/r/gnoswap/gnft/gnft_test.gno

```diff
@@ -0,0 +1,450 @@
+package gnft
+
+import (
+	"std"
+	"testing"
+
+	"gno.land/p/demo/grc/grc721"
+	"gno.land/p/demo/testutils"
+	"gno.land/p/demo/uassert"
+
+	"gno.land/r/gnoswap/v1/consts"
+)
+
+const (
+	errInvalidTokenId = "invalid token id"
+)
+
+var (
+	positionAddr  = consts.POSITION_ADDR
+	positionRealm = std.NewCodeRealm(consts.POSITION_PATH)
+
+	addr01      = testutils.TestAddress("addr01")
+	addr01Realm = std.NewUserRealm(addr01)
+
+	addr02      = testutils.TestAddress("addr02")
+	addr02Realm = std.NewUserRealm(addr02)
+)
+
+func TestMetadata(t *testing.T) {
+	tests := []struct {
+		name     string
+		fn       func() string
+		expected string
+	}{
+		{"Name()", Name, "GNOSWAP NFT"},
+		{"Symbol()", Symbol, "GNFT"},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			uassert.Equal(t, tt.expected, tt.fn())
+		})
+	}
+}
+
+func TestTotalSupply(t *testing.T) {
+	tests := []struct {
+		name     string
+		setup    func()
+		expected uint64
+	}{
+		{
+			name:     "initial total supply",
+			expected: uint64(0),
+		},
+		{
+			name: "total supply after minting",
+			setup: func() {
+				std.TestSetRealm(positionRealm)
+				Mint(a2u(addr01), tid(1))
+			},
+			expected: uint64(1),
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if tt.setup != nil {
+				tt.setup()
+			}
+			uassert.Equal(t, tt.expected, TotalSupply())
+		})
+	}
+}
+
+func TestBalanceOf(t *testing.T) {
+	tests := []struct {
+		name     string
+		addr     std.Address
+		expected uint64
+	}{
+		{"BalanceOf(addr01)", addr01, uint64(1)},
+		{"BalanceOf(addr02)", addr02, uint64(0)},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			uassert.Equal(t, tt.expected, BalanceOf(a2u(tt.addr)))
+		})
+	}
+}
+
+func TestOwnerOf(t *testing.T) {
+	tests := []struct {
+		name        string
+		tokenId     uint64
+		shouldPanic bool
+		panicMsg    string
+		expected    std.Address
+	}{
+		{"OwnerOf(1)", 1, false, "", addr01},
+		{"OwnerOf(2)", 2, true, errInvalidTokenId},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if tt.shouldPanic {
+				uassert.PanicsWithMessage(t, tt.panicMsg, func() {
+					OwnerOf(tid(tt.tokenId))
+				})
+			} else {
+				uassert.Equal(t, tt.expected, OwnerOf(tid(tt.tokenId)))
+			}
+		})
+	}
+}
+
+func TestIsApprovedForAll(t *testing.T) {
+	tests := []struct {
+		name     string
+		setup    func()
+		expected bool
+	}{
+		{
+			name:     "IsApprovedForAll(addr01, addr02)",
+			expected: false,
+		},
+		{
+			name: "IsApprovedForAll(addr01, addr02) after setting approval",
+			setup: func() {
+				std.TestSetRealm(addr01Realm)
+				SetApprovalForAll(a2u(addr02), true)
+			},
```

Comment:
An additional TC is required to verify that the asset can be transferred from the Approved address.

Reply:
2e25c41db0c6c37365eed4918f2ad05521b8c61f